main_test.go

```diff
@@ -203,56 +204,188 @@ func TestController(t *testing.T) {
+							Labels: map[string]string{
+								"a":              "b",
+								hashringLabelKey: "hashring0",
+							},
+						},
+						Spec: appsv1.StatefulSetSpec{
+							Replicas:    intPointer(3),
+							ServiceName: "h0",
+						},
+					},
+				})
 
 			// Reconciliation is async, so we need to wait a bit.
 			<-time.After(500 * time.Millisecond)
-			cm, err = klient.CoreV1().ConfigMaps(opts.namespace).Get(opts.configMapGeneratedName, metav1.GetOptions{})
+			gcm, err := klient.CoreV1().ConfigMaps(opts.namespace).Get(opts.configMapGeneratedName, metav1.GetOptions{})
 			if err != nil {
 				t.Fatalf("got unexpected error getting ConfigMap: %v", err)
 			}
 
-			buf, err = json.Marshal(expected)
+			// Manually change properties of generated configmap.
+			gcm.Labels = labels
+			if _, err = klient.CoreV1().ConfigMaps(opts.namespace).Update(gcm); err != nil {
+				t.Fatalf("got unexpected error updating ConfigMap: %v", err)
+			}
+
+			buf, err := json.Marshal(hashrings)
 			if err != nil {
-				t.Fatalf("got unexpected error marshaling expected hashrings: %v", err)
+				t.Fatalf("got unexpected error marshaling initial hashrings: %v", err)
+			}
+			cm.Data = map[string]string{
+				opts.fileName: string(buf),
+			}
+			if cm, err = klient.CoreV1().ConfigMaps(opts.namespace).Update(cm); err != nil {
+				t.Fatalf("got unexpected error updating ConfigMap: %v", err)
 			}
 
-			if cm.Data[opts.fileName] != string(buf) {
-				t.Errorf("the expected config does not match the actual config\ncase:\t%q\ngiven:\t%+v\nexpected:\t%+v\n", name, cm.Data[opts.fileName], string(buf))
+			// Reconciliation is async, so we need to wait a bit.
+			<-time.After(500 * time.Millisecond)
+			gcm, err = klient.CoreV1().ConfigMaps(opts.namespace).Get(opts.configMapGeneratedName, metav1.GetOptions{})
+			if err != nil {
+				t.Fatalf("got unexpected error getting ConfigMap: %v", err)
+			}
+
+			if shouldBeUpdated {
+				// Generated configmap should be overfidden by default properties.
+				if cmp.Equal(gcm.Labels, map[string]string{}) {
+					t.Errorf("generated configmap should have been updated\ncase:\t%q\noriginal:\t%+v\ngenerated:\t%+v\n", name, cm.Labels, gcm.Labels)
+				}
+			} else {
+				// Generated configmap should preserve changes.
+				if !cmp.Equal(gcm.Labels, labels) {
+					t.Errorf("generated configmap should NOT have been updated\ncase:\t%q\noriginal:\t%+v\ngenerated:\t%+v\n", name, cm.Labels, gcm.Labels)
+				}
 			}
 		})
 	}
 }
+
+func setup(t *testing.T, klient kubernetes.Interface, opts *options) func() {
+	c := newController(klient, nil, opts)
+	stop := make(chan struct{})
+
+	//nolint:staticcheck
+	go func() {
+		if err := c.run(stop); err != nil {
+			t.Fatalf("got unexpected error: %v", err)
+		}
+	}()
+
+	return func() {
+		close(stop)
+	}
+}
+
+func createOriginalConfigMap(t *testing.T, klient kubernetes.Interface, opts *options, hashrings []receive.HashringConfig, statefulsets []*appsv1.StatefulSet) *corev1.ConfigMap {
```

Comment:
this also creates statefulsets. consider naming: `createInitialResources` or `bootstrapResources`